server/src/main/java/org/elasticsearch/search/vectors/VectorData.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.vectors;
+
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HexFormat;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.common.Strings.format;
+
+public record VectorData(float[] floatVector, byte[] byteVector) implements Writeable, ToXContentFragment {
+
+    public static final String XCONTENT_PARAM_NAME = "field";
+    private static final String DEFAULT_XCONTENT_NAME = "vector";
+
+    private VectorData(float[] floatVector) {
+        this(floatVector, null);
+    }
+
+    private VectorData(byte[] byteVector) {
+        this(null, byteVector);
+    }
+
+    public VectorData(StreamInput in) throws IOException {
+        this(in.readOptionalFloatArray(), in.readOptionalByteArray());
+    }
+
+    public VectorData {
+        if (false == (floatVector == null ^ byteVector == null)) {
+            throw new IllegalArgumentException("please supply exactly either a float or a byte vector");
+        }
+    }
+
+    public byte[] asByteVector() {
+        if (byteVector != null) {
+            return byteVector;
+        }
+        DenseVectorFieldMapper.ElementType.BYTE.checkVectorBounds(floatVector);
+        byte[] vec = new byte[floatVector.length];
+        for (int i = 0; i < floatVector.length; i++) {
+            vec[i] = (byte) floatVector[i];
+        }
+        return vec;
+    }
+
+    // Explicit byte parsing is only supported when operation on hex-encoded byte vectors
+    // which are defined only for ElementType.BYTE, so we should throw if we request to convert it to float
+    public float[] asFloatVector() {
+        return asFloatVector(true);
+    }
+
+    public float[] asFloatVector(boolean failIfByte) {
```

Comment:
Don't really like this - as we only need to force the conversion for serialization.  Maybe it'd be better to have a separate  (albeit very similar) method instead, or include this logic somehow in the `writeTo` ? The main challenge there is that in certain cases (e.g. `KnnVectorQueryBuilder`) there is additional logic in-between handling the `query_vector` param.

Reply:
Thinking more on it, I don't think we should do this. Right now we allow "byte" arrays to query float indexed values.

Consider the following:

 - New Coordinator accepts the `hex` encoded byte array
 - Its serialized to an older node (thus transformed to `float[]`)
 - Then we successfully query a `element_type: float` field.

I think its ok for us to be lenient here. 

What do you think @mayya-sharipova. It seems that preventing byte[] array queries against a `element_type: float` field is causing more trouble than its worth :(

Reply:
Yeah, I don't think we should fail here. `float v = b` is valid in java as its expanding the values. It seems like an unnecessary restriction the more I think about it. 

Sorry for flip-flopping on this so much. What do you think @pmpailis should we restrict it? @mayya-sharipova what say you? I am happy to go with the majority as I obviously cannot make up my mind :)

Reply:
Currently we do support `byte` values for `float` fields (we won't know about element type until much later) so I guess it makes sense to be "consistent" for hex as well and let it pass for `float` vectors. I do understand the reasoning for restricting this, but we don't do that for a standard byte array now either, so this could potentially cause some confusion.  

\+ the scenario you mentioned, unless we decide to throw if `hex` & old nodes (which I don't really think it'd be nice), would make us be lenient :)

Reply:
Updated (temporarily) to **not** fail for byte -> float conversion. Happy to change back if we decide to do so :) @mayya-sharipova wdyt?